lib/before-watchPatterns.js

```diff
@@ -1,16 +1,33 @@
-const { AppDirectoryLocation } = require("./constants");
+const { basename } = require("path");
+const { buildEnvData, getCompilationContext } = require("./utils");
 
 module.exports = function (hookArgs) {
-	if (hookArgs.liveSyncData && hookArgs.liveSyncData.bundle) {
-		return (args, originalMethod) => {
-			return originalMethod(...args).then(originalPatterns => {
-				const appDirectoryLocationIndex = originalPatterns.indexOf(AppDirectoryLocation);
-				if (appDirectoryLocationIndex !== -1) {
-					originalPatterns.splice(appDirectoryLocationIndex, 1);
-				}
+    const { liveSyncData } = hookArgs;
+    if (!liveSyncData || !liveSyncData.bundle) {
+        return;
+    }
 
-				return originalPatterns;
-			});
-		};
-	}
+    const { platforms } = hookArgs;
+    const { env } = liveSyncData;
+    return (args, originalMethod) => {
+        return originalMethod(...args).then(originalPatterns => {
+            if (!platforms.length) {
+                throw new Error("Target platform should be specified!");
+            }
+
+            const compilationContexts = platforms.map(platform =>
+                getContext(platform, env));
+
+            const ignorePatterns = compilationContexts.map(
+                context => `!${context}`
+            );
+
+            return [...originalPatterns, ...ignorePatterns];
```

Comment:
The old logic `splice`d the unneeded watch pattern, whereas the new logic adds an exclusion and retains the `originalPatterns`.
Whenever we have `["app", "app/App_Resources", "!app"]` in watch patterns - does this work as expected?

Reply:
Discussed in person.